### PR TITLE
[JENKINS-46207] Enable tag discovery

### PR DESCRIPTION
**Pull request by stephenc**
- I suspect there are additional changes required before can merge for git
- I suspect there are additional changes required to branch-api before safe to merge for users

See [JENKINS-46207](https://issues.jenkins-ci.org/browse/JENKINS-46207)

@reviewbybees

### CODE REVIEW

**Comment by stephenc** [comment]
w00t! I do believe this is complete support for the Git plugin:

You can configure discovery of tags:

configuring tag discovery

And tags are discovered, but not automatically built (unless you install some plugin that implements a [BranchBuildStrategy](https://github.com/jenkinsci/branch-api-plugin/blob/master/src/main/java/jenkins/branch/BranchBuildStrategy.java) that builds tags and configure your job with that strategy

discovered tags

File: src/main/java/jenkins/plugins/git/traits/BranchDiscoveryTrait.java [hunk 1]
```diff
@@ -146,6 +148,14 @@ public String getDisplayName() {
             public boolean isApplicableToOrigin(@NonNull Class<? extends SCMHeadOrigin> originClass) {
                 return SCMHeadOrigin.Default.class.isAssignableFrom(originClass);
             }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public boolean isApplicableToHead(@NonNull Class<? extends SCMHeadMixin> headClass) {
+                return super.isApplicableToHead(headClass) && !(TagSCMHead.class.isAssignableFrom(headClass));
```
Comment (stephenc) [inline src/main/java/jenkins/plugins/git/traits/BranchDiscoveryTrait.java]: for historical reasons, because Git plugin does not implement its own SCMHead for branches, we have to use this more complex check to exclude this authority from being applied to tags

File: src/test/java/jenkins/plugins/git/AbstractGitSCMSourceTest.java [hunk 2]
```diff
@@ -96,6 +98,112 @@ public void retrieveHeadsRequiresBranchDiscovery() throws Exception {
...
+        TaskListener listener = StreamTaskListener.fromStderr();
+        // SCMHeadObserver.Collector.result is a TreeMap so order is predictable:
+        assertEquals("[]", source.fetch(listener).toString());
+        source.setTraits(Collections.<SCMSourceTrait>singletonList(new TagDiscoveryTrait()));
+        assertEquals("[SCMHead{'annotated'}, SCMHead{'lightweight'}]", source.fetch(listener).toString());
+        // And reuse cache:
+        assertEquals("[SCMHead{'annotated'}, SCMHead{'lightweight'}]", source.fetch(listener).toString());
+    }
+
+    @Issue("JENKINS-45953")
```
Comment (stephenc) [inline src/test/java/jenkins/plugins/git/AbstractGitSCMSourceTest.java]: Strictly an unrelated change, but as JENKINS-45953 is IMHO not a bug but a configuration being enforced, I have added this test now that we can correctly discover tags and hence have `fetchRevisions` return values based on what is discoverable

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 3]
```diff
@@ -355,19 +357,30 @@ protected GitTool resolveGitTool(String gitTool) {
...
+                                          if (branchName.equals(head.getName())) {
+                                              return new SCMRevisionImpl(head, b.getSHA1String());
+                                          }
                                       }
                                   }
                                   return null;
                               }
                           },
-                new GitSCMSourceContext<>(null, SCMHeadObserver.none()).withTraits(getTraits()),
+                context,
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: Why was this moved up? AFAICT you are not using this variable until now.

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 4]
```diff
@@ -509,6 +522,112 @@ public void record(@NonNull SCMHead head, SCMRevision revision, boolean isMatch)
...
+                                            }
+                                        }
+                                    };
+                                }
+                            }, new SCMSourceRequest.LazyRevisionLambda<GitTagSCMHead, GitTagSCMRevision, ObjectId>() {
+                                @NonNull
+                                @Override
+                                public GitTagSCMRevision create(@NonNull GitTagSCMHead head, @Nullable ObjectId intermediate)
+                                        throws IOException, InterruptedException {
+                                    return new GitTagSCMRevision(head, ref.getValue().name());
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: wow we really need `-source 8`

@MarkEWaite BTW [the great majority of people updating this plugin are already on a version of Jenkins requiring Java 8](http://stats.jenkins.io/pluginversions/git.html) :-)

File: src/main/java/jenkins/plugins/git/GitTagSCMRevision.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class GitTagSCMRevision extends AbstractGitSCMSource.SCMRevisionImpl {
+    /**
+     * Constructor.
+     *
+     * @param head the head.
+     * @param hash the revision hash.
+     */
+    public GitTagSCMRevision(@NonNull GitTagSCMHead head, @NonNull String hash) {
+        super(head, hash);
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitTagSCMRevision.java]: Do we need to override `getHead` to be type-safe?

**Comment by chris5287** [comment]
Any update on this? The review seems to have gone quiet :(

File: src/main/java/jenkins/plugins/git/GitTagSCMRevision.java (hunk 5 above)
Comment (stephenc) [reply src/main/java/jenkins/plugins/git/GitTagSCMRevision.java]: getHead is final.

**Comment by stephenc** [comment]
@chris5287 I've been moved to a different focus, I'm pinging internally to see if we can get progress

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java (hunk 3 above)
Comment (stephenc) [reply src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: May have been a left-over of some earlier refactoring where I was filtering the revision discovery based on the context, e.g. I may have had

```
if (context.wantTags() && head instanceof GitTagSCMHead) {
...
} else if (context.wantBranches()) {
...
}
```

within the closure... but then I decided that conflicted. If you have a valid `SCMHead`, even if it would not be discovered by the `fetch(TaskListener)` we should still report its revision (as long as it exists)

**Comment by michaelneale** [comment]
can someone kick the build off again? I think that was a kill due to memory, or something.

**Comment by stephenc** [comment]
@michaelneale just close and reopen the PR to trigger a rebuild (I've just done that BTW)

**Comment by stephenc** [comment]
So it is working on all Windows builds and at least one linux build... and it builds on my machine... seems like just the CI server playing maggots

screen shot 2017-09-25 at 12 29 10

**Comment by michaelneale** [comment]
@stephen https://github.com/jenkinsci/git-plugin/blob/master/Jenkinsfile#L9 means that one of the linux builds failing isn't linux failing, but a version of jenkins. I am not sure what `null` means in that context, perhaps the latest one, so it may be a legitimate failure with a certain version.

**Comment by michaelneale** [comment]
`[linux-8] Cancelling nested steps due to timeout`

**Comment by chris5287** [comment]
@michaelneale @stephenc LTS is now 2.73.1, does need updating?

**Comment by michaelneale** [comment]
@chris5287 https://github.com/jenkinsci/git-plugin/pull/533  - see what that one does?

**Comment by michaelneale** [comment]
@stephenc this is failing on https://github.com/jenkinsci/git-plugin/blob/master/pom.xml#L27 - version of Jenkins, but passing on a recent LTS. should the pom.xml be updated to a newer version?

**Comment by chris5287** [comment]
The current io.jenkins.archetypes:plugin outputs `<jenkins.version>2.7.3</jenkins.version>` (https://github.com/jenkinsci/archetypes/blob/6b4d30543457456333715b760bb1124946c6bbb3/hello-world/src/main/resources/archetype-resources/pom.xml#L15) and I assume the `<java.level>7</java.level>` is wrong now that Jenkins is Java 8 only (https://jenkins.io/doc/book/installing/#system-requirements)?

**Comment by MarkEWaite** [comment]
@chris5287 the plugin intentionally remains based on Jenkins 1.625.3 (the first to require Java 7) until it is ready to transition to Java 8.  When it transitions to Java 8, it will be done with a change of major version number and will depend on Jenkins 2.60.3 or 2.73.2.

The java.level needs to remain at 7 so that the code continues to run on older Java 7 based environments.

I expect the switch to Java 8 to happen before end of calendar 2017.

**Comment by chris5287** [comment]
@MarkEWaite thanks for the explanation. Any idea when a new release is scheduled for this plugin?

**Comment by MarkEWaite** [comment]
Likely this weekend or early next week.

**Comment by weakcamel** [comment]
Sounds like release of this feature has been rescheduled which is of course fair.
Any chance for new ETA please? :-)

**Comment by MarkEWaite** [comment]
It was released with [git plugin 3.6.0](https://wiki.jenkins.io/display/JENKINS/Git+Plugin#GitPlugin-ChangeLog) 2 Oct 2017.

**Comment by cleverkidd** [comment]
hm, for me it partially works (= does not work)
I got refspec `+refs/tags/*:refs/remotes/origin/tags/*`
1) it fetches the repository: `git fetch --tags --progress origin +refs/tags/*:refs/remotes/origin/tags/*`
`git ls-remote <REPO_URL>`
2) few other steps
3) it wants to remove objects that were removed at remote repo: `git fetch --tags --progress origin +refs/tags/*:refs/remotes/origin/tags/* --prune`
`git ls-remote -h <REPO_URL>`

in step 3 option -h will retrieve `heads` only, which is wrong

**Comment by stephenc** [comment]
@cleverkidd could you please move this to the Jenkins issue tracker. Please provide there the configuration of your multibranch project (sanitised as necessary) and whether this is an issue affecting: 1. discovery; 2. build; 3. both. The log file where you see this sequence of events (sanitised and trimmed to stop post checkout as necessary) will also be required to identify.

FTR if you have added the "Discover Tags" trait then the [only place where GitSCMSource calls ls-remote](https://github.com/jenkinsci/git-plugin/blob/837ae6424c7636b6f7aaabbe8337ae3a2c053d0a/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L408) will also add the `-t` 

There is a second call in the SCMSource side, [when fallback mode is trying to determine the master branch](https://github.com/jenkinsci/git-plugin/blob/837ae6424c7636b6f7aaabbe8337ae3a2c053d0a/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L727) but that never adds `-h`

That is the only two cases where branch indexing could call `git ls-remote`

From what I can see there is only [one case](https://github.com/jenkinsci/git-plugin/blob/c1ac7012d4003a1bf2394031cb1a76baac016523/src/main/java/hudson/plugins/git/GitSCM.java#L633) where building a specific tag could trigger a `git ls-remote`. That will have a `-h` and should be immediately followed by `Found __N__ remote heads on __url__` *but* that should only be invoked during polling, and multibranch does not do polling

**Comment by stephenc** [comment]
> @cleverkidd could you please move this to the Jenkins issue tracker. 

Just to be clear. Create a new JIRA issue, component `git-plugin`

**Comment by ramesbury** [comment]
This feature is not available with in a Github Organization
image

**Comment by stephenc** [comment]
There is a PR for GitHub tag discovery against the GitHub plugin
